core/store/src/lib.rs

```diff
@@ -187,9 +189,34 @@ impl StoreUpdate {
     }
 
     pub fn commit(self) -> Result<(), io::Error> {
+        debug_assert!(
+            self.transaction.ops.len()
+                == self
+                    .transaction
+                    .ops
+                    .iter()
+                    .map(|op| match op {
+                        DBOp::Insert { col, key, .. } => {
+                            (*col as u8, key)
+                        }
+                        DBOp::Delete { col, key } => {
+                            (*col as u8, key)
+                        }
+                    })
+                    .collect::<std::collections::HashSet<_>>()
+                    .len(),
+            "Transaction overwrites itself: {:?}",
+            self
+        );
         if let Some(trie) = self.trie {
             assert_eq!(
-                trie.storage.as_caching_storage().unwrap().store.storage.as_ref() as *const _,
+                trie.get_trie_for_shard(0)
```

Comment:
Is it going to work with many shards?

Reply:
I guess it should because storage is the same